File: dart/utils/btree_printer.dart
part of tree_node;

class BTreePrinter {

  static void printNode(TreeNode root) {
    int maxLevel = BTreePrinter.maxLevel(root);
    BTreePrinter.printNodeInternal([root], 1, maxLevel);
  }

  static void printNodeInternal(List<TreeNode> nodes, int level, int maxLevel) {
    if (nodes.isEmpty || BTreePrinter.isAllElementsNull(nodes)) {
      return;
    }

    StringBuffer sb = new StringBuffer();
    int floor = maxLevel - level;
    int endgeLines = pow(2, max(floor - 1, 0)).toInt(); // prob not needed
    int firstSpaces = (pow(2, floor) - 1).toInt();  // prob not needed
    int betweenSpaces = (pow(2, (floor + 1)) - 1).toInt();

    BTreePrinter.printWhitespaces(firstSpaces, sb);

    List<TreeNode> newNodes = new List<TreeNode>();
    for (TreeNode node in nodes) {
      if (node != null) {
        sb.write(node.data);
        newNodes.add(node.left);
        newNodes.add(node.right);
      } else {
        newNodes.add(null);
        newNodes.add(null);
        sb.write(" ");
      }


      BTreePrinter.printWhitespaces(betweenSpaces, sb);
    }

    print(sb.toString());
    sb.clear();

    for (int i = 1; i <= endgeLines; i++) {
      for (int j = 0; j < nodes.length; j++) {
        BTreePrinter.printWhitespaces(firstSpaces - i, sb);
        if (nodes[j] == null) {
          BTreePrinter.printWhitespaces(endgeLines + endgeLines + i + 1, sb);
          continue;
        }

        if (nodes[j].left != null) {
          sb.write("/");
        } else {
          BTreePrinter.printWhitespaces(1, sb);
        }

        BTreePrinter.printWhitespaces(i + i - 1, sb);

        if (nodes[j].right != null) {
          sb.write("\\");
        } else {
          BTreePrinter.printWhitespaces(1, sb);
        }

        BTreePrinter.printWhitespaces(endgeLines + endgeLines - i, sb);
      }

      print(sb.toString());
      sb.clear();
    }

    BTreePrinter.printNodeInternal(newNodes, level + 1, maxLevel);
  }

  static void printWhitespaces(int count, StringBuffer sb) {
    for (int i = 0; i < count; i++) {
      sb.write(" ");
    }
  }

  static int maxLevel(TreeNode node) {
    if (node == null) {
      return 0;
    }

    return max(BTreePrinter.maxLevel(node.left), BTreePrinter.maxLevel(node.right)) + 1;
  }

  static bool isAllElementsNull(List list) {
    for (var i in list) {
      if (i != null) {
        return false;
      }
    }

    return true;
  }
}

File: c-sharp/ctci.Library/BTreePrinter.cs
using System;
using System.Collections.Generic;

namespace ctci.Library
{
    public class BTreePrinter
    {
        public static void PrintNode(TreeNode root) {
            int maxLevel = BTreePrinter.MaxLevel(root);

            PrintNodeInternal(new List<TreeNode>() { root }, 1, maxLevel);
        }

        private static void PrintNodeInternal(List<TreeNode> nodes, int level, int maxLevel) {
            if (nodes.Count==0 || BTreePrinter.IsAllElementsNull(nodes))
                return;

            int floor = maxLevel - level;
            int endgeLines = (int) Math.Pow(2, (Math.Max(floor - 1, 0)));
            int firstSpaces = (int) Math.Pow(2, (floor)) - 1;
            int betweenSpaces = (int) Math.Pow(2, (floor + 1)) - 1;

            BTreePrinter.PrintWhitespaces(firstSpaces);

            List<TreeNode> newNodes = new List<TreeNode>();
            foreach (TreeNode node in nodes) {
                if (node != null) {
                    Console.Write(node.Data);
                    newNodes.Add(node.Left);
                    newNodes.Add(node.Right);
                } else {
                    newNodes.Add(null);
                    newNodes.Add(null);
                    Console.Write(" ");
                }

                BTreePrinter.PrintWhitespaces(betweenSpaces);
            }
            Console.WriteLine();

            for (int i = 1; i <= endgeLines; i++) {
                for (int j = 0; j < nodes.Count; j++) {
                    BTreePrinter.PrintWhitespaces(firstSpaces - i);
                    if (nodes[j] == null) {
                        BTreePrinter.PrintWhitespaces(endgeLines + endgeLines + i + 1);
                        continue;
                    }

                    if (nodes[j].Left != null)
                        Console.Write("/");
                    else
                        BTreePrinter.PrintWhitespaces(1);

                    BTreePrinter.PrintWhitespaces(i + i - 1);

                    if (nodes[j].Right != null)
                        Console.Write("\\");
                    else
                        BTreePrinter.PrintWhitespaces(1);

                    BTreePrinter.PrintWhitespaces(endgeLines + endgeLines - i);
                }

                Console.WriteLine();
            }

            PrintNodeInternal(newNodes, level + 1, maxLevel);
        }

        private static void PrintWhitespaces(int count)
        {
            string padding = string.Format("{0}", count);
            padding = "{0," + padding + "}";
            Console.Write(padding, " ");
            //for (int i = 0; i < count; i++)
            //    Console.Write(" ");
        }

        private static int MaxLevel(TreeNode node) {
            if (node == null)
                return 0;

            return Math.Max(BTreePrinter.MaxLevel(node.Left), BTreePrinter.MaxLevel(node.Right)) + 1;
        }

        private static bool IsAllElementsNull<T>(IEnumerable<T> list) {
            foreach (object o in list) {
                if (o != null)
                    return false;
            }

            return true;
        }
    }
}
